// @framer/clayton.primitive-button/code/Button.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import styled from "styled-components";

// Define type of property
interface Props {
  styleSet: string;
  disabledButton: boolean;
  disabledOpacity: number;
  fullHeight: boolean;
  fullWidth: boolean;
  label: string;
  fontSize: number;
  fontWeight: string;
  lineHeight: number;
  letterSpacing: number;
  customFont: boolean;
  customFontFamily: string;
  textColor: string;
  textColorHover: string;
  textColorFocus: string;
  textAlignment: string;
  wrapText: boolean;
  icons: string;
  leftIconArt: string;
  leftIconArtHover: string;
  leftIconSize: number;
  leftIconOffset: number;
  rightIconArt: string;
  rightIconArtHover: string;
  rightIconSize: number;
  rightIconOffset: number;
  paddingPerFontSize: boolean;
  paddingPerSide: boolean;
  paddingNaturalFactor: number;
  padding: number;
  paddingTop: number;
  paddingRight: number;
  paddingBottom: number;
  paddingLeft: number;
  borderWidth: number;
  borderColor: string;
  borderColorHover: string;
  borderColorFocus: string;
  borderRadius: number;
  useBackground: boolean;
  bgColor: string;
  bgColorHover: string;
  bgColorFocus: string;
  boxShadow: boolean;
  boxShadowColor: string;
  boxShadowBlur: number;
  boxShadowOffset: number;
  boxShadowSpread: number;
  transitionSpeed: number;
  tabIndex: number;
}

export class Button extends React.Component<Props> {

  // Set default properties
  static defaultProps = {
    // UI Control Groups
    styleSet: "1",
    // Disabled?
    disabledButton: false,
    disabledOpacity: .5,
    // Size
    width: 200,
    height: 48,
    fullWidth: true,
    fullHeight: true,
    // Text
    label: "Button Label",
    fontSize: 16,
    fontWeight: '400',
    lineHeight: 1,
    letterSpacing: 0,
    customFont: false,
    textColor: "#000",
    textColorHover: "#FFF",
    textColorFocus: "#FFF",
    textAlignment: "center",
    wrapText: false,
    // Icon
    icons: "No",
    leftIconArt: "",
    leftIconArtHover: "",
    leftIconSize: .4,
    leftIconOffset: 16,
    rightIconArt: "",
    rightIconArtHover: "",
    rightIconSize: .4,
    rightIconOffset: 16,
    // Padding
    paddingPerFontSize: true,
    paddingPerSide: true,
    paddingNaturalFactor: 1,
    padding: 16,
    paddingTop: 16,
    paddingRight: 32,
    paddingBottom: 16,
    paddingLeft: 32,
    // Border
    borderWidth: 1,
    borderColor: "rgba(0,0,0,0.2)",
    borderColorHover: "#09F",
    borderColorFocus: "#06F",
    borderRadius: 2,
    // Background
    useBackground: true,
    bgColor: "#FFF",
    bgColorHover: "#09F",
    bgColorFocus: "#06F",
    // Background
    boxShadow: false,
    boxShadowColor: "rgba(0,0,0,.2)",
    boxShadowBlur: 5,
    boxShadowOffset: 3,
    boxShadowSpread: 0,
    // Transition
    transitionSpeed: .2,
    // Misc
    tabIndex: 0
  }

  static propertyControls: PropertyControls<Props> = {
    disabledButton: {
      type: ControlType.Boolean,
      title: "Disabled",
      enabledTitle: "Yes",
      disabledTitle: "No",
    },
    disabledOpacity: {
      type: ControlType.Number,
      title: ".   opacity",
      min: 0,
      max: 1,
      step: .1,
      hidden: (props) => props.disabledButton == false
    },
    styleSet: {
      type: ControlType.Enum,
      title: "Settings",
      options: ["1", "2", "3"],
      optionTitles: ["** Label **", "** Button **", "** Icons **"],
    },
    label: {
      type: ControlType.String,
      title: "Label",
      hidden: (props) => props.styleSet !== "1"
    },
    fontSize: {
      type: ControlType.Number,
      title: ".   size",
      min: 1,
      unit: "px",
      hidden: (props) => props.styleSet !== "1"
    },
    fontWeight: {
      type: ControlType.Enum,
      title: ".   weight",
      options: ["100", "200", "300", "400", "500", "600", "700", "800", "900"],
      // optionTitles: ["Thin", "Extra Light", "Light", "Normal", "Medium", "Semi Bold", "Bold", "Extra Bold", "Heavy"]
      hidden: (props) => props.styleSet !== "1"
    },
    textAlignment: {
      type: ControlType.SegmentedEnum,
      title: ".   alignment",
      options: ["flex-start", "center", "flex-end"],
      optionTitles: ["L", "C", "R"],
    },
    lineHeight: {
      type: ControlType.Number,
      title: ".   leading",
      min: 0,
      max: 10,
      step: .1,
      hidden: (props) => props.styleSet !== "1"
    },
    letterSpacing: {
      type: ControlType.Number,
      title: ".   spacing",
      min: 0,
      max: 10,
      step: .1,
      unit: "px",
      hidden: (props) => props.styleSet !== "1"
    },
    textColor: {
      type: ControlType.Color,
      title: ".   default",
      hidden: (props) => props.styleSet !== "1"
    },
    textColorHover: {
      type: ControlType.Color,
      title: ".   hover",
      hidden: (props) => props.styleSet !== "1"
    },
    textColorFocus: {
      type: ControlType.Color,
      title: ".   focus",
      hidden: (props) => props.styleSet !== "1"
    },
    wrapText: {
      type: ControlType.Boolean,
      title: ".   wrap",
      enabledTitle: "Yes",
      disabledTitle: "No",
      hidden: (props) => props.styleSet !== "1"
    },
    customFont: {
      type: ControlType.Boolean,
      title: ".   font",
      enabledTitle: "Custom",
      disabledTitle: "System",
      hidden: (props) => props.styleSet !== "1"
    },
    customFontFamily: {
      type: ControlType.String,
      title: ".   fontstack",
      hidden: (props) => props.styleSet !== "1" || props.customFont == false
    },
    fullWidth: {
      type: ControlType.Boolean,
      title: "Width",
      enabledTitle: "100%",
      disabledTitle: "Natural",
      hidden: (props) => props.styleSet !== "2"
    },
    fullHeight: {
      type: ControlType.Boolean,
      title: "Height",
      enabledTitle: "100%",
      disabledTitle: "Natural",
      hidden: (props) => props.styleSet !== "2"
    },
    paddingPerFontSize: {
      type: ControlType.Boolean,
      title: "Padding",
      enabledTitle: 'Font Size',
      disabledTitle: 'Literal',
      hidden: (props) => props.styleSet !== "2" || props.fullWidth == true && props.fullHeight == true
    },
    paddingNaturalFactor: {
      type: ControlType.Number,
      title: ".   factor",
      step: .1,
      unit: "em",
      hidden: (props) => props.styleSet !== "2" || props.paddingPerFontSize == false || props.fullWidth == true && props.fullHeight == true
    },
    padding: {
      type: ControlType.FusedNumber,
      toggleKey: "paddingPerSide",
      toggleTitles: ["All Sides", "Per Side"],
      valueKeys: ["paddingTop", "paddingRight", "paddingBottom", "paddingLeft"],
      valueLabels: ["T", "R", "B", "L"],
      min: 0,
      title: ".   offsets",
      hidden: (props) => props.styleSet !== "2" || props.paddingPerFontSize == true || props.fullWidth == true && props.fullHeight == true
    },
    borderWidth: {
      type: ControlType.Number,
      title: "Border",
      min: 0,
      max: 6,
      displayStepper: true,
      unit: "px",
      hidden: (props) => props.styleSet !== "2"
    },
    borderColor: {
      type: ControlType.Color,
      title: ".   default",
      hidden: (props) => props.styleSet !== "2" || props.borderWidth <= 0
    },
    borderColorHover: {
      type: ControlType.Color,
      title: ".   hover",
      hidden: (props) => props.styleSet !== "2" || props.borderWidth <= 0
    },
    borderColorFocus: {
      type: ControlType.Color,
      title: ".   focus",
      hidden: (props) => props.styleSet !== "2" || props.borderWidth <= 0
    },
    borderRadius: {
      type: ControlType.Number,
      title: ".   radius",
      min: 0,
      displayStepper: true,
      unit: "px",
      hidden: (props) => props.styleSet !== "2"
    },
    useBackground: {
      type: ControlType.Boolean,
      title: "Background",
      enabledTitle: "Yes",
      disabledTitle: "No",
      hidden: (props) => props.styleSet !== "2"
    },
    bgColor: {
      type: ControlType.Color,
      title: ".   default",
      hidden: (props) => props.styleSet !== "2" || props.useBackground == false
    },
    bgColorHover: {
      type: ControlType.Color,
      title: ".   hover",
      hidden: (props) => props.styleSet !== "2" || props.useBackground == false
    },
    bgColorFocus: {
      type: ControlType.Color,
      title: ".   focus",
      hidden: (props) => props.styleSet !== "2" || props.useBackground == false
    },
    transitionSpeed: {
      type: ControlType.Number,
      title: "Transition",
      min: 0,
      max: 6,
      step: 0.1,
      unit: "s",
      hidden: (props) => props.styleSet !== "2"
    },
    boxShadow: {
      type: ControlType.Boolean,
      title: "Shadow",
      enabledTitle: "Yes",
      disabledTitle: "No",
      hidden: (props) => props.styleSet !== "2"
    },
    boxShadowColor: {
      type: ControlType.Color,
      title: ".   color",
      hidden: (props) => props.styleSet !== "2" || props.boxShadow == false
    },
    boxShadowBlur: {
      type: ControlType.Number,
      title: ".   blur",
      min: 0,
      max: 100,
      unit: "px",
      hidden: (props) => props.styleSet !== "2" || props.boxShadow == false
    },
    boxShadowOffset: {
      type: ControlType.Number,
      title: ".   offset",
      min: -50,
      max: 50,
      unit: "px",
      hidden: (props) => props.styleSet !== "2" || props.boxShadow == false
    },
    boxShadowSpread: {
      type: ControlType.Number,
      title: ".   spread",
      min: -50,
      max: 50,
      unit: "px",
      hidden: (props) => props.styleSet !== "2" || props.boxShadow == false
    },
    tabIndex: {
      type: ControlType.Number,
      title: "Tab Index",
      displayStepper: true,
      hidden: (props) => props.styleSet !== "2"
    },
    icons: {
      type: ControlType.SegmentedEnum,
      title: "Icons",
      options: ["No", "L", "R", "LR"],
      hidden: (props) => props.styleSet !== "3"
    },
    leftIconArt: {
      type: ControlType.File,
      allowedFileTypes: ["png", "jpg", "svg"],
      title: ".   L art",
      hidden: (props) => props.styleSet !== "3" || props.icons == "No" || props.icons == "R"
    },
    leftIconArtHover: {
      type: ControlType.File,
      allowedFileTypes: ["png", "jpg", "svg"],
      title: ".   L hover",
      hidden: (props) => props.styleSet !== "3" || props.icons == "No" || props.icons == "R"
    },
    leftIconSize: {
      type: ControlType.Number,
      title: ".   L size",
      min: 0,
      max: 5,
      step: .1,
      hidden: (props) => props.styleSet !== "3" || props.icons == "No" || props.icons == "R"
    },
    leftIconOffset: {
      type: ControlType.Number,
      title: ".   L offset",
      min: 0,
      hidden: (props) => props.styleSet !== "3" || props.icons == "No" || props.icons == "L"
    },
    rightIconArt: {
      type: ControlType.File,
      allowedFileTypes: ["png", "jpg", "svg"],
      title: ".   R art",
      hidden: (props) => props.styleSet !== "3" || props.icons == "No" || props.icons == "L"
    },
    rightIconArtHover: {
      type: ControlType.File,
      allowedFileTypes: ["png", "jpg", "svg"],
      title: ".   R hover",
      hidden: (props) => props.styleSet !== "3" || props.icons == "No" || props.icons == "R"
    },
    rightIconSize: {
      type: ControlType.Number,
      title: ".   R size",
      min: 0,
      max: 5,
      step: .1,
      hidden: (props) => props.styleSet !== "3" || props.icons == "No" || props.icons == "L"
    },
    rightIconOffset: {
      type: ControlType.Number,
      title: ".   R offset",
      min: 0,
      hidden: (props) => props.styleSet !== "3" || props.icons == "No" || props.icons == "L"
    },
  };

  StyledButton = styled.button`
        display: flex;
        flex-direction: row;
        align-items: center;
        box-sizing: border-box;
        border: none;
        margin: 0;
        outline: none;
        transition: all ${this.props.transitionSpeed}s;
        &:hover {
            color: ${this.props.textColorHover} !important;
            box-shadow: inset 0 0 0 ${this.props.borderWidth}px ${this.props.borderColorHover} !important;
            background-color: ${this.props.useBackground ? this.props.bgColorHover : 'transparent'} !important;
        }
        &:focus {
            color: ${this.props.textColorFocus} !important;
            box-shadow: inset 0 0 0 ${this.props.borderWidth}px ${this.props.borderColorFocus} !important;
            background-color: ${this.props.useBackground ? this.props.bgColorFocus : 'transparent'} !important;
        }
        &:hover, &:focus {
            .button-icon-left {
                background-image: ${this.props.leftIconArtHover ? "url('" + this.props.leftIconArtHover + "')" : "url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iOCIgaGVpZ2h0PSIxNCIgdmlld0JveD0iMCAwIDggMTQiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CjxwYXRoIGQ9Ik03IDEzTDEgN0w3IDEiIHN0cm9rZT0iYmxhY2siIHN0cm9rZS13aWR0aD0iMiIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIi8+Cjwvc3ZnPgo=')"} !important;
            }
            .button-icon-right {
                background-image: ${this.props.rightIconArtHover ? "url('" + this.props.rightIconArtHover + "')" : "url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iOCIgaGVpZ2h0PSIxNCIgdmlld0JveD0iMCAwIDggMTQiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CjxwYXRoIGQ9Ik0xIDEzTDcgN0wxIDEiIHN0cm9rZT0iYmxhY2siIHN0cm9rZS13aWR0aD0iMiIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIi8+Cjwvc3ZnPgo=')"} !important;
            }
        }
        &[disabled] {
            pointer-events: none;
            opacity: ${this.props.disabledOpacity};
        }
    `
  StyledIcon = styled.span`
        align-self: stretch;
        background-repeat: no-repeat;
        background-position: 50% 50%;
    `

  render() {
    const {
      fullHeight,
      disabledButton,
      paddingPerFontSize,
      paddingPerSide,
      paddingNaturalFactor,
      padding,
      paddingTop,
      paddingRight,
      paddingBottom,
      paddingLeft,
      letterSpacing,
      tabIndex,
    } = this.props;

    const paddingTopValue =
      fullHeight ? '0' : (paddingPerFontSize ? `${paddingNaturalFactor}em` : (paddingPerSide ? `${paddingTop}px` : `${padding}px`))
    const paddingBottomValue =
      fullHeight ? '0' : (paddingPerFontSize ? `${paddingNaturalFactor}em` : (paddingPerSide ? `${paddingBottom}px` : `${padding}px`))
    const paddingRightValue =
      paddingPerFontSize ? `${paddingNaturalFactor * 2}em` : (paddingPerSide ? `${paddingRight}px` : `${padding}px`)
    const paddingLeftValue =
      paddingPerFontSize ? `${paddingNaturalFactor * 2}em` : (paddingPerSide ? `${paddingLeft}px` : `${padding}px`)
    const paddingValue = `${paddingTopValue} ${paddingRightValue} ${paddingBottomValue} ${paddingLeftValue}`

    let leftIcon;
    let rightIcon;

    if (this.props.icons == "L" || this.props.icons == "LR") {
      leftIcon = <this.StyledIcon className="button-icon-left" style={{
        width: this.props.leftIconSize + "em",
        marginRight: this.props.leftIconOffset,
        backgroundImage: this.props.leftIconArt ? "url('" + this.props.leftIconArt + "')" : "url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iOCIgaGVpZ2h0PSIxNCIgdmlld0JveD0iMCAwIDggMTQiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CjxwYXRoIGQ9Ik03IDEzTDEgN0w3IDEiIHN0cm9rZT0iYmxhY2siIHN0cm9rZS13aWR0aD0iMiIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIi8+Cjwvc3ZnPgo=')",
        backgroundSize: this.props.leftIconSize + "em auto",
      }} />
    }
    if (this.props.icons == "R" || this.props.icons == "LR") {
      rightIcon = <this.StyledIcon className="button-icon-right" style={{
        width: this.props.rightIconSize + "em",
        marginLeft: this.props.rightIconOffset,
        backgroundImage: this.props.rightIconArt ? "url('" + this.props.rightIconArt + "')" : "url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iOCIgaGVpZ2h0PSIxNCIgdmlld0JveD0iMCAwIDggMTQiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CjxwYXRoIGQ9Ik0xIDEzTDcgN0wxIDEiIHN0cm9rZT0iYmxhY2siIHN0cm9rZS13aWR0aD0iMiIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIi8+Cjwvc3ZnPgo=')",
        backgroundSize: this.props.rightIconSize + "em auto",
      }} />
    }

    return (
      <this.StyledButton
        style={{
          width: this.props.fullWidth ? "100%" : "auto",
          height: this.props.fullHeight ? "100%" : "auto",
          fontSize: this.props.fontSize,
          fontWeight: this.props.fontWeight,
          fontFamily: this.props.customFont ? this.props.customFontFamily : "font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
          whiteSpace: this.props.wrapText ? "normal" : "nowrap",
          lineHeight: this.props.lineHeight,
          letterSpacing: letterSpacing > 0 ? `${this.props.letterSpacing}px` : null,
          justifyContent: this.props.textAlignment,
          color: this.props.textColor,
          padding: paddingValue,
          borderRadius: this.props.borderRadius,
          backgroundColor: this.props.useBackground ? this.props.bgColor : "transparent",
          boxShadow: "inset 0 0 0 " + this.props.borderWidth + "px " + this.props.borderColor + (this.props.boxShadow ? ", 0 " + this.props.boxShadowOffset + "px " + this.props.boxShadowBlur + "px " + this.props.boxShadowSpread + "px " + this.props.boxShadowColor : ""),
        }}
        disabled={disabledButton ? disabledButton : disabledButton}
        tabIndex={tabIndex}
      >
        {leftIcon}
        {this.props.label}
        {rightIcon}
      </this.StyledButton>
    );
  }
}
